corehq/apps/users/models_role.py

```diff
@@ -257,6 +262,7 @@ def as_permission_info(self):
         return PermissionInfo(self.permission, allow=allow)
 
 
+@audit_fields("value")
```

Comment:
I'm curious how this will work. These models are only ever created (when we add a new permission) or deleted (if we remove a permission from the platform). Both of those are quite rare and have corresponding code changes to the `Permissions` model. 

From an auditing point of view, how will the data for this model be used?

Reply:
The `RolePermission` model maps `UserRole`s to these permissions (via the `permission_fk` field), so auditing these values is needed in order to know _how_ a `RolePermission` change affects the `UserRole` that it references. Although I just noticed that I missed the `id` field, I'll add that.

Reply:
I think you misunderstand how these work. There are only currently 38 rows in this table on any env. These are reference values that essentially never change. We do not create / remove these as roles are created / edited.

The only time we will ever see an audit event for this model is when we create a new permission in the system via a Django migration.

Reply:
Yes, I understand that. An audit event record _at that time_ is important because it will record the primary key of the `SQLPermission` that was changed.

Reply:
> SQLPermission that was changed

Can you point me to code that changes the value of a SQLPermission?

Reply:
When I say "changed" I mean any of add/delete/update.  These are all changes that need to be audited. SQLPermission records were changed at least [here](https://github.com/dimagi/commcare-hq/blob/de94bebd971/corehq/apps/users/migrations/0043_add_release_management_permission.py#L7) and [here](https://github.com/dimagi/commcare-hq/blob/de94bebd971/corehq/apps/users/migrations/0045_add_view_tableau_permission.py#L7).

Reply:
So what's the value of having audit records for these creates? I still don't understand how this will ever be useful.

Reply:
The `RolePermission` model has `role` and `permission_fk` foreign keys that link a `UserRole` to a `SQLPermission`.  An audit event for a `RolePermission` change only documents the foreign key values for these fields. Maintaining audit records that document what a `SQLPermission` _means_ (i.e. the properties of the permission itself and which `id` they refer to) is crucial in understanding (and historically retaining) the impact of a `RolePermission` change.

Reply:
So if you are looking at an audit record for the `RolePermission` are you saying that you'll look up the value of `SQLPermission` in the audit data and not in the primary DB?

Since these records are simple values I would have thought it makes more logical sense to just record the value in the audit record instead of the FK.

Reply:
> are you saying that you'll look up the value of SQLPermission in the audit data and not in the primary DB?

Yes, because from a "historical audit" perspective, there is no guarantee that the primary DB value exists (or has the same value) as it did when the audit event occurred.

Reply:
So what you're suggesting is that for every single audit row that has a foreign key you'd need to query the audit data for the latest foreign key model prior to the date of the audit row. And if the foreign key model has no row in the audit data then you'll fall back to the main table. Is that correct?

Reply:
> you'd need to query the audit data for the latest foreign key model prior to the date of the audit row

Correct, this is how you would know what the value was at the time an audit event occurred.

> if the foreign key model has no row in the audit data then you'll fall back to the main table

No, if the FK model doesn't have an audit data record, you cannot know what its value was at that time, period. Using the current value in the primary table would be a guess.

Reply:
> No, if the FK model doesn't have an audit data record, you cannot know what its value was at that time, period. Using the current value in the primary table would be a guess.

Yes, you can know. It's pretty logical really. If there is no value in the audit data then the value has not changed since auditing began (unless it has been changed outside of the purview of the audit framework in which case all bets are off anyway).

Since we're talking about permissions here, the data for those is already created so if you're saying it's impossible to know what the value was if it doesn't existing in the audit data then how are we going to get that data for all the existing permissions? Seems to suggest that adding auditing for this model is pointless. Unless you're planning to bootstrap the audit data with all the existing data?

Reply:
> Unless you're planning to bootstrap the audit data with all the existing data?

Yes, I've referred to this as "backfilling", and [the TODO item for this feature](https://github.com/dimagi/django-field-audit/blob/dev/README.md?plain=1#L156) is a very high priority. I've even considered adding it before merging _this_ PR. I'm still undecided on this.

Reply:
> Happy to go to voice on this.

Sorry, I don't know how I missed this when you said it earlier. I'd be happy to cover this in the Auditability meeting this week.